### ISSUE TITLE
Add content: typing

### ISSUE THREAD

**Issue by louisguitton** [koaning/calmcode-feedback#7]
First of all big congrats on the content there. I didn't know any of your sites, but you pointed me to this one on a PR I opened on whatlies and I binge watched some of the content.

I know, not very calm.

But the format resonated with me. And the list of available tools too.

One tool that I feel could be in the natural progression of flake8, black, pre-commit etc... is IMHO `mypy`.
Ideas for chapters:
-  `the problem`
    simple intro to type annotations in a `code.py`. You could include something like this
    ```python
    def hello(name, adjective = None):
        return f"Hello {adjective} {name}"
    ```
    this doesn't have types to start with, then you would do
    ```python
    def hello(name: str, adjective: str = None) -> str:
        return f"Hello {adjective} {name}"
    ```
    this would fail with `error: Argument 2 of "hello" has incompatible type "Optional[str]"; expected "str"` which you could fix with 
    ```python
    def hello(name: str, adjective: str = "") -> str:
        return f"Hello {adjective} {name}"
    ```
- installing mypy and running `mypy code.py`
- ignoring errors on one line with `# type: ignore`
- supressing errors for third party libraries 
    ```
    ; mypy.ini
    [mypy]
    ignore_missing_imports = True
    ```
- mention the typing cheatsheet https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html or https://www.pythonsheets.com/notes/python-typing.html

Other things maybe worth mentioning:
- mypy is not perfect: it has 1000+ open issues, and is lagging behind sometimes. E.g. if you use dataclasses with functions: mypy will break with a false positive `error: Too many arguments for ...`

**Comment by koaning** [comment]
I am very happy you're mentioning this since`mypy` definately is on the shortlist of tools I'd like to add. The autocompletion benefits are too awesome to ignore (especially when you like method chaining ... and I *really* do). I'd also love to add some `pydantic` after it (check it out if you haven't [here](https://pydantic-docs.helpmanual.io/)) . 

To keep the stress at bay though I try to make content "when I feel like" and "what I feel like at the moment". At the moment that I'll do [annoy](https://github.com/spotify/annoy) and [schedule](https://schedule.readthedocs.io/en/stable/) first. There's a good chance that I want to to do the pandas/scikit-learn stuff too soon. Especially with [scikit-lego](https://scikit-lego.readthedocs.io/en/latest/) and [scikit-fairness](https://github.com/koaning/scikit-fairness) being projects I intend on maintaining.

All in all, enough content in the pipeline. But `mypy` will get attention, hopefully soon-ish. 

Also, great to hear you're enjoying the content and I appreciate you sharing some resources here. I agree that `mypy` has many benefits and that it may not be something to add everywhere at all times but it certainly has improved a lot of code-bases that I have worked with. 

I'll keep the issue open until the content is up.

**Comment by koaning** [comment]
I figured I should mention it since you may find it relevant (you're also in NLP), I'm employed by [Rasa](https://rasa.com/docs/) and on behalf of them I also support a youtube playlist on NLP algorithms [here](https://www.youtube.com/watch?v=wWNMST6t1TA&list=PL75e0qA87dlG-za8eLI6t0_Pbxafk-cxb). If there's any content suggestions for that channel I'd also be all ears.

**Comment by koaning** [comment]
Need to check if this tweet is true, otherwise it might be a nice example of mypy not being perfect https://twitter.com/gail_w/status/1269985482601050118.

**Comment by louisguitton** [comment]
Ah yeah. If you want to enforce type annotations (and are already running flake8), you can use something like https://github.com/python-discord/flake8-annotations

**Comment by koaning** [comment]
I'm wondering if instead of putting the focus on `mypy`... wouldn't it be better to focus on typing?

**Comment by louisguitton** [comment]
Yeah! actually this is quite broad. You could talk rather about "type validation" which is basically what you do over on the fastapi series. Or you could talk about "static checking" which is what you touch upon with black and pre-commit.

Just for reference, I'll put the 3 typing related europython talks here :
- [Static typing in python](https://ep2020.europython.eu/talks/9BZ96n9-static-typing-in-python/) (amazing intro, could be abused for this issue)
- [Fullstack type safety](https://ep2020.europython.eu/talks/5muMqs3-full-stack-type-safety/) (was great in a sense that it showed how SQL is typed while python and JS are loose, so you then have to make JS less loose with typescript and python less loose with mypy and ORms)
- [tooling-for-static-analysis-of-python-programs](https://ep2020.europython.eu/talks/3k93S3j-tooling-for-static-analysis-of-python-programs/) (quite advanced talk on how to parse AST trees better using his tools)
